bulldozer/signals.go

```diff
@@ -21,109 +21,389 @@ import (
+	if err != nil {
+		return false, "", err
+	}
+
+	_, matches.PRBodySubstrings, err = s.prBodyMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	_, matches.Branches, err = s.branchMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	_, matches.BranchPatterns, err = s.branchPatternMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	_, matches.MaxCommits, err = s.maxCommitsMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	logger := zerolog.Ctx(ctx)
+	logger.Debug().Msgf("matches count: %d", matches.size())
+	logger.Debug().Msgf("signals count: %d", s.size())
+
+	if matches.size() == s.size() {
+		return true, fmt.Sprintf("pull request matches all %s signals", tag), nil
+	}
+
+	return false, fmt.Sprintf("pull request does not match all %s signals", tag), nil
+}
+
+// MatchesAny returns true if the pull request meets one or more signals. It also
 // returns a description of the signal that was met. The tag argument appears
 // in this description and indicates the behavior (trigger, ignore) this
 // set of signals is associated with.
-func (s *Signals) Matches(ctx context.Context, pullCtx pull.Context, tag string) (bool, string, error) {
+func (s *Signals) MatchesAny(ctx context.Context, pullCtx pull.Context, tag string) (bool, string, error) {
+	descriptions, matches, err := s.labelMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	if len(matches) > 0 {
+		return true, descriptions[0], nil
+	}
+
+	descriptions, matches, err = s.commentMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	if len(matches) > 0 {
+		return true, descriptions[0], nil
+	}
+
+	descriptions, matches, err = s.commentSubstringMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	if len(matches) > 0 {
+		return true, descriptions[0], nil
+	}
+
+	descriptions, matches, err = s.prBodyMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	if len(matches) > 0 {
+		return true, descriptions[0], nil
+	}
+
+	descriptions, matches, err = s.branchMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	if len(matches) > 0 {
+		return true, descriptions[0], nil
+	}
+
+	descriptions, matches, err = s.branchPatternMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	if len(matches) > 0 {
+		return true, descriptions[0], nil
+	}
+
+	return false, fmt.Sprintf("pull request does not match the %s", tag), nil
+}
+
+// labelMatches determines which label signals match the given PR. It returns:
+// - A list of descriptions for each matched signal
+//   - These will only include the first item the signal matched when there is more than one
+// - A list of the matched signals
+func (s *Signals) labelMatches(ctx context.Context, pullCtx pull.Context, tag string) ([]string, Signal, error) {
```

Comment:
Following on from my comment above about all/any, this signature feels a bit weird:

* Multiple descriptions are returned but they're either all discarded or only the first is used
* Multiple matching signals are returned, but ultimately callers only care about the count

Going into this, I was expecting these to have the `(bool, string, error)` signature of the callers, which admittedly only works if the "all" behavior applies when combining conditions, not within a single condition.

If we do keep the current behavior, `(int, string, error)` might be a better signature.